libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -522,91 +522,102 @@ bool ContractCompiler::visit(InlineAssembly const& _inlineAssembly)
-							);
-						for (unsigned i = 0; i < variable->type()->sizeOnStack(); ++i)
-							_assembly.append(dupInstruction(stackDiff));
+						solAssert(location.second == 0, "Intra-slot offest assumed to be zero.");
+						_assembly.append(location.first);
 					}
 					else
 					{
-						solAssert(m_context.isStateVariable(variable), "Invalid variable type.");
-						auto const& location = m_context.storageLocationOfVariable(*variable);
-						if (!variable->type()->isValueType())
-						{
-							solAssert(location.second == 0, "Intra-slot offest assumed to be zero.");
-							_assembly.append(location.first);
-						}
-						else
-						{
-							_assembly.append(location.first);
-							_assembly.append(u256(location.second));
-						}
+						_assembly.append(location.first);
+						_assembly.append(u256(location.second));
 					}
 				}
-				else if (auto contract = dynamic_cast<ContractDefinition const*>(decl))
-				{
-					solAssert(contract->isLibrary(), "");
-					_assembly.appendLibraryAddress(contract->fullyQualifiedName());
-				}
-				else
-					solAssert(false, "Invalid declaration type.");
-			} else {
-				// lvalue context
-				auto variable = dynamic_cast<VariableDeclaration const*>(decl);
-				solAssert(
-					!!variable && m_context.isLocalVariable(variable),
-					"Can only assign to stack variables in inline assembly."
+			}
+			else if (auto contract = dynamic_cast<ContractDefinition const*>(decl))
+			{
+				solAssert(contract->isLibrary(), "");
+				_assembly.appendLibraryAddress(contract->fullyQualifiedName());
+			}
+			else
+				solAssert(false, "Invalid declaration type.");
+		}
+		else
+		{
+			// lvalue context
+			auto variable = dynamic_cast<VariableDeclaration const*>(decl);
+			solAssert(
+				!!variable && m_context.isLocalVariable(variable),
+				"Can only assign to stack variables in inline assembly."
+			);
+			unsigned size = variable->type()->sizeOnStack();
+			int stackDiff = _assembly.deposit() - m_context.baseStackOffsetOfVariable(*variable) - size;
+			if (stackDiff > 16 || stackDiff < 1)
+				BOOST_THROW_EXCEPTION(
+					CompilerError() <<
+					errinfo_sourceLocation(_inlineAssembly.location()) <<
+					errinfo_comment("Stack too deep, try removing local variables.")
 				);
-				unsigned size = variable->type()->sizeOnStack();
-				int stackDiff = _assembly.deposit() - m_context.baseStackOffsetOfVariable(*variable) - size;
-				if (stackDiff > 16 || stackDiff < 1)
-					BOOST_THROW_EXCEPTION(
-						CompilerError() <<
-						errinfo_sourceLocation(_inlineAssembly.location()) <<
-						errinfo_comment("Stack too deep, try removing local variables.")
-					);
-				for (unsigned i = 0; i < size; ++i) {
-					_assembly.append(swapInstruction(stackDiff));
-					_assembly.append(Instruction::POP);
-				}
+			for (unsigned i = 0; i < size; ++i)
+			{
+				_assembly.append(swapInstruction(stackDiff));
+				_assembly.append(Instruction::POP);
```

Comment:
* [ ] I also need to check that `<expression>` leaves at least one stack element
* [ ] I need to check that in `x := <expression>`, `x :=` part comes immediately after `<expression>`.
* [ ] I need to identify named regions in this context.  Is it a contiguous region on the stack or more than one contiguous regions?
* [ ] I need to check that code desugared from the strict mode cannot have `=: x` that touches the named regions

> functional statements that do not leave items on the stack

If that kind of functional statements exist, you can use `x := <functional statement that do not leave items on the stack>` to dig into the stack.  Is this correct?